File: Rust/src/solution/add_to_array_form.rs
use crate::solution::Solution;

impl Solution {
    pub fn add_to_array_form(num: Vec<i32>, k: i32) -> Vec<i32> {
        let mut result = Vec::new();
        let mut carry = k;
        for i in (0..num.len()).rev() {
            let sum = num[i] + carry;
            result.push(sum % 10);
            carry = sum / 10;
        }
        while carry > 0 {
            result.push(carry % 10);
            carry /= 10;
        }
        result.reverse();
        result
    }
}

File: Dart/lib/addToArrayForm.dart
class Solution {
  List<int> addToArrayForm(List<int> num, int k) {
    List<int> result = [];
    int carry = 0;
    int i = num.length - 1;
    while ((i >= 0) || (k > 0) || (carry > 0)) {
      int sum = carry;
      if (i >= 0) {
        sum += num[i];
        i--;
      }
      if (k > 0) {
        sum += k % 10;
        k ~/= 10;
      }
      result.add(sum % 10);
      carry = sum ~/ 10;
    }
    return result.reversed.toList();
  }
}
